internal/stacks/stackmigrate/load.go

```diff
@@ -0,0 +1,163 @@
+	// based on what we had from the `terraform init` command.
+	var backend backend.Backend
+	var backendConfig cty.Value
+
+	// the absence of backend state file indicates a local backend
+	if backendState == nil {
+		backend = local.New()
+		backendConfig = cty.ObjectVal(map[string]cty.Value{
+			"path":          cty.StringVal(fmt.Sprintf("%s/%s", configPath, "terraform.tfstate")),
+			"workspace_dir": cty.StringVal(configPath),
+		})
+	} else {
+		initFn := backendInit.Backend(backendState.Backend.Type)
+		if initFn == nil {
+			diags = diags.Append(fmt.Errorf("unknown backend type %q", backendState.Backend.Type))
+			return state, diags
+		}
+
+		backend = initFn()
+		schema := backend.ConfigSchema()
+		config, err := backendState.Backend.Config(schema)
+		if err != nil {
+			diags = diags.Append(tfdiags.Sourceless(
+				tfdiags.Error,
+				"Failed to decode current backend config",
+				fmt.Sprintf("The backend configuration created by the most recent run of \"terraform init\" could not be decoded: %s. The configuration may have been initialized by an earlier version that used an incompatible configuration structure. Run \"terraform init -reconfigure\" to force re-initialization of the backend.", err),
+			))
+			return state, diags
+		}
+
+		var moreDiags tfdiags.Diagnostics
+		backendConfig, moreDiags = backend.PrepareConfig(config)
+		diags = diags.Append(moreDiags)
+		if moreDiags.HasErrors() {
+			return state, diags
+		}
+
+		// it's safe to ignore terraform version conflict between the local and remote environments,
+		// as we are only reading the state
+		if backendR, ok := backend.(*remote.Remote); ok {
+			backendR.IgnoreVersionConflict()
+		}
+	}
+
+	// Now that we have the backend and its configuration, we can configure it.
+	moreDiags := backend.Configure(backendConfig)
+	diags = diags.Append(moreDiags)
+	if moreDiags.HasErrors() {
+		return state, diags
+	}
+
+	// The backend is initialised and configured, so now we can load the state
+	// from the backend.
+	stateManager, err := backend.StateMgr(workspace)
+	if err != nil {
+		diags = diags.Append(fmt.Errorf("error loading state: %s", err))
+		return state, diags
+	}
+
+	// We'll lock the backend here to ensure that we don't have any concurrent
+	// operations on the state. If this fails, we'll return an error and the
+	// user should retry the migration later when nothing is currently updating
+	// the state.
+	id, err := stateManager.Lock(statemgr.NewLockInfo())
+	if err != nil {
+		diags = diags.Append(tfdiags.Sourceless(tfdiags.Error, "Failed to lock state", fmt.Sprintf("The state is currently locked by another operation: %s. Please retry the migration later.", err)))
+		return state, diags
+	}
+	if err := stateManager.RefreshState(); err != nil {
+		diags = diags.Append(fmt.Errorf("error loading state: %s", err))
+		return state, diags
+	}
+	state = stateManager.State()
+
+	// Remember to unlock the state when we're done.
+	if err := stateManager.Unlock(id); err != nil {
```

Comment:
I think this should be in a defer statement? For example, line 133 will return without unlocking the state.